sample: dfu: allow sample to be build on default platform

jfischer-no opened the pull request:
Remove platform_allow and allow sample to be build on
default platform but not on native posix.

jfischer-no committed:
sample: dfu: remove platform_allow

Remove platform_allow and allow sample to be build on
compatible platform only.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

de-nordic commented:
You probably want to exclude all platforms that do not define `image_0` partition.

jfischer-no committed:
twister: add dt_compat_enabled_with_label DT filter function

The function returns true if the node with the referenced
node label is compatible and enabled.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

carlescufi commented:
@de-nordic and @PerMac could you please take a look?

de-nordic commented:
@jfischer-no seems that the builtkite problems, at least with the mimxrt1064_evk, have existed prior to the PR but have been uncovered here.
I have tried to build the sample from master (f0352500b3e5c3aaecca17d03b1e07bb2e0ac973) and found that I can not even go through configuration phase to menuconfig.
Question is whether the fix for these problems should be part of this PR, or the problems should be filtered out and fixed as part of other PR.

de-nordic commented:
@jfischer-no Is this still relevant?

de-nordic reviewed (approved):
Looks OK

de-nordic commented:
@PerMac Can you also look at this?

PerMac commented:
Looks good. Just please change `sanitycheck:` to `twister:` in the commit title as there is no more "sanitycheck" :)

carlescufi commented:
@galak and @mbolivar-nordic please take a look

mbolivar-nordic commented on scripts/pylib/twister/expr_parser.py:
```diff
@@ -264,6 +264,13 @@ def ast_expr(ast, env, edt):
+    elif ast[0] == "dt_compat_enabled_with_label":
+        compat = ast[1][0]
+        label = ast[1][1]
+        for node in edt.nodes:
+            if node.status == "okay" and label in node.labels and node.matching_compat == compat:
+                return True
+        return False
```
I'm late to the party, but this should use `edt.label2node` instead for speed. Something like this (untested):

```py
node = edt.label2node.get(label)
return node is not None and node.status == 'okay' and node.matching_compat == compat
```

@jfischer-no can you send a follow-up?

jfischer-no replied:
@mbolivar-nordic thanks for the hint, see #33387
